builtins/src/cost_modeling.rs

```diff
@@ -0,0 +1,300 @@
+//! Configurations for handling cost modeling of builtin programs.
+
+use {
+    crate::BUILTINS,
+    ahash::AHashMap,
+    lazy_static::lazy_static,
+    solana_feature_set::FeatureSet,
+    solana_pubkey::Pubkey,
+    solana_sdk_ids::{ed25519_program, secp256k1_program},
+};
+
+// Hehe, if you change me, I will break things...
+pub const NUM_COST_MODELED_BUILTINS_WITH_MIGRATIONS: usize = 3;
+
+/// Configuration for cost modeling of a builtin program.
+#[derive(Debug)]
+pub enum CostModelingConfig {
+    /// The builtin program is cost modeled.
+    CostModeled {
+        /// The default cost of the builtin program.
+        default_cost: u64,
+    },
+    /// The builtin program is not cost modeled.
+    NotCostModeled,
+}
+
+impl CostModelingConfig {
+    /// Returns `true` if the builtin program is cost modeled.
+    pub fn is_cost_modeled(&self) -> bool {
+        matches!(self, Self::CostModeled { .. })
+    }
+}
+
+struct BuiltinCost {
+    default_cost: u64,
+    core_bpf_migration_feature: Option<Pubkey>,
+}
+
+#[derive(Copy, Clone, Default)]
+struct BuiltinWithMigration {
+    program_id: Pubkey,
+    feature_id: Pubkey,
+}
+
+lazy_static! {
+    static ref BUILTIN_INSTRUCTION_COSTS: AHashMap<Pubkey, BuiltinCost> = BUILTINS
```

Comment:
a note for future refactors to remove `BUILTIN_INSTRUCTION_COSTS`: this `AHashMap` has two functions: 
1. CostModel uses it to determine if program is builtin. It can be done by checking if `NotCostModeled`
2. Provides lookup by `program_id`. Since the list of builtins is small and static, could implement customized lookup instead.

Reply:
Another note, would like to profile perf between `AHashmap::get()` and `binary_search` if `BUILTINS` is sorted,

Reply:
Okay, I can profile this. Afaict `BUILTINS` being sorted shouldn't break anything in the runtime.